pkg/runtime/runtime.go

```diff
@@ -2909,14 +2990,80 @@ func (a *DaprRuntime) getAppHTTPChannelConfig(pipeline httpMiddleware.Pipeline,
 		MaxRequestBodySizeMB: a.runtimeConfig.maxRequestBodySize,
 	}
 
-	if !isExternal {
-		conf.Endpoint = a.getAppHTTPEndpoint()
-		conf.Client = a.appHTTPClient
-	} else {
-		conf.Client = nethttp.DefaultClient
+	conf.Endpoint = a.getAppHTTPEndpoint()
+	conf.Client = a.appHTTPClient
+	return conf
+}
+
+func (a *DaprRuntime) getHTTPEndpointAppChannel(pipeline httpMiddleware.Pipeline, endpoint httpEndpointV1alpha1.HTTPEndpoint) (httpChannel.ChannelConfiguration, error) {
+	conf := httpChannel.ChannelConfiguration{
+		CompStore:            a.compStore,
+		MaxConcurrency:       a.runtimeConfig.appConnectionConfig.MaxConcurrency,
+		Pipeline:             pipeline,
+		MaxRequestBodySizeMB: a.runtimeConfig.maxRequestBodySize,
 	}
 
-	return conf
+	if a.globalConfig.Spec.TracingSpec != nil {
+		conf.TracingSpec = *a.globalConfig.Spec.TracingSpec
+	}
+
+	var tlsConfig *tls.Config
+
+	if endpoint.HasTLSRootCA() {
+		ca := endpoint.Spec.ClientTLS.RootCA.Value.String()
+		caCertPool := x509.NewCertPool()
+
+		if !caCertPool.AppendCertsFromPEM([]byte(ca)) {
+			return httpChannel.ChannelConfiguration{}, fmt.Errorf("failed to add root cert to cert pool for http endpoint %s", endpoint.ObjectMeta.Name)
+		}
+
+		tlsConfig = &tls.Config{
+			MinVersion: tls.VersionTLS12,
+			RootCAs:    caCertPool,
+		}
+	}
+
+	if endpoint.HasTLSPrivateKey() {
+		cert, err := tls.X509KeyPair([]byte(endpoint.Spec.ClientTLS.Certificate.Value.String()), []byte(endpoint.Spec.ClientTLS.PrivateKey.Value.String()))
+		if err != nil {
+			return httpChannel.ChannelConfiguration{}, fmt.Errorf("failed to load client certificate for http endpoint %s: %w", endpoint.ObjectMeta.Name, err)
+		}
+
+		if tlsConfig == nil {
+			tlsConfig = &tls.Config{MinVersion: tls.VersionTLS12}
+		}
+		tlsConfig.Certificates = []tls.Certificate{cert}
+	}
+
+	if endpoint.Spec.ClientTLS != nil && endpoint.Spec.ClientTLS.Renegotiation != nil {
+		switch *endpoint.Spec.ClientTLS.Renegotiation {
+		case commonapi.NegotiateNever:
+			tlsConfig.Renegotiation = tls.RenegotiateNever
+		case commonapi.NegotiateOnceAsClient:
+			tlsConfig.Renegotiation = tls.RenegotiateOnceAsClient
+		case commonapi.NegotiateFreelyAsClient:
+			tlsConfig.Renegotiation = tls.RenegotiateFreelyAsClient
+		default:
+			return httpChannel.ChannelConfiguration{}, fmt.Errorf("invalid renegotiation value %s for http endpoint %s", *endpoint.Spec.ClientTLS.Renegotiation, endpoint.ObjectMeta.Name)
+		}
+	}
+
+	dialer := &net.Dialer{
+		Timeout: 15 * time.Second,
+	}
+
+	netTransport := &nethttp.Transport{
```

Comment:
It may be better to clone `http.DefaultTransport` and modify the cloned copy rather than creating one ex-novo. https://stackoverflow.com/a/66018065

Reply:
Changed